drivers/char/tpm/tpm-interface.c

```diff
@@ -953,6 +953,11 @@ int tpm_do_selftest(struct tpm_chip *chip)
 	loops = jiffies_to_msecs(duration) / delay_msec;
 
 	rc = tpm_continue_selftest(chip);
+	if (rc == TPM_ERR_INVALID_POSTINIT) {
+		chip->flags |= TPM_CHIP_FLAG_ALWAYS_POWERED;
+		if (tpm_startup(chip))
+			return rc;
```

Comment:
This `if` clause has no effect: when `tpm_startup` fails (< 0) execution will follow the branch and run `return rc`, but when it succeeds (0) the next check will be `if (rc)`, which is true because `rc` is `TPM_ERR_INVALID_POSTINIT`, so `return rc` will run in the very same way.

If we decide to pursue this option (manually starting the TPM) I believe you should call `rc = tpm_continue_selftest(chip);` to re-run the TPM self-test, update `rc`, and let the remainder of the function run and perform the remaining tests.

That being said, it seems safer to remove the call to `tpm_startup` as well and only set the flag, leaving it inoperable instead of trying to recover from the this error. This should be enough to avoid the suspend failure, and it looks like it is also the preferred approach upstream: https://lists.gt.net/linux/kernel/1359824#1359824